src/targets/ref/lowering.cpp

```diff
@@ -307,20 +307,46 @@ struct ref_quant_gemm
     {
         argument result{output_shape};
         // first, convert the args[0] and args[1] from int8_t to int32_t
-        argument arg_0{{shape::int32_type, {args.at(0).get_shape().lens()}}};
-        argument arg_1{{shape::int32_type, {args.at(1).get_shape().lens()}}};
-        arg_0.visit([&](auto output) {
-            args.at(0).visit(
-                [&](auto input) { std::copy(input.begin(), input.end(), output.begin()); });
-        });
-
-        arg_1.visit([&](auto output) {
-            args.at(1).visit(
-                [&](auto input) { std::copy(input.begin(), input.end(), output.begin()); });
-        });
+        argument arg_0{{output_shape.type(), {args.at(0).get_shape().lens()}}};
+        argument arg_1{{output_shape.type(), {args.at(1).get_shape().lens()}}};
+        if(output_shape.type() == migraphx::shape::float_type)
+        {
+            arg_0.visit([&](auto output) {
+                args.at(0).visit([&](auto input) {
+                    std::transform(input.begin(), input.end(), output.begin(), [&](const auto x) {
+                        return static_cast<float>(x);
+                    });
+                });
+            });
 
-        migemm(result, arg_0, arg_1, int32_t{1}, int32_t{0});
+            arg_1.visit([&](auto output) {
+                args.at(1).visit([&](auto input) {
+                    std::transform(input.begin(), input.end(), output.begin(), [&](const auto x) {
+                        return static_cast<float>(x);
+                    });
+                });
+            });
+            migemm(result, arg_0, arg_1, 1.0f, 0.0f);
+        }
+        else if(output_shape.type() == migraphx::shape::int32_type)
+        {
+            arg_0.visit([&](auto output) {
+                args.at(0).visit([&](auto input) {
+                    std::transform(input.begin(), input.end(), output.begin(), [&](const auto x) {
+                        return static_cast<int32_t>(x);
+                    });
+                });
+            });
 
+            arg_1.visit([&](auto output) {
+                args.at(1).visit([&](auto input) {
+                    std::transform(input.begin(), input.end(), output.begin(), [&](const auto x) {
+                        return static_cast<int32_t>(x);
+                    });
+                });
+            });
```

Comment:
We shouldn't be converting the input here. [`gemm`](https://github.com/ROCm/AMDMIGraphX/blob/develop/src/include/migraphx/gemm.hpp#L52) is already using a `double` accumulator. We could also convert the `amat` and `bmat` to double to avoid loss with the multiply as well.

Reply:
Done.